### PR TITLE
[s-1]-Implement-registration-for-the-service

### PR DESCRIPTION
**Pull request by ValentynaBondarenko**
-add migration schema User;
-add entity User;
-implement validation methods for name,emaile,pass;
-implement tests for validator;
-add UserService;

### CODE REVIEW

File: src/main/java/com/bondarenko/controller/UserController.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+@Slf4j
+@RestController
+public class UserController {
+    @Autowired
+    private UserService userService;
+
+    @ResponseStatus(HttpStatus.CREATED)
+    @PostMapping("/signup")
+    public String registrationUser(User user) {
```
Comment (chickitosik) [inline src/main/java/com/bondarenko/controller/UserController.java:19]: use dto and RequestBody

File: src/main/java/com/bondarenko/dto/user/AuthenticationRequest.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    }
+
+    public String getUserName() {
+        return userName;
+    }
+
+    public void setUserName(String userName) {
+        this.userName = userName;
+    }
+
```
Comment (chickitosik) [inline src/main/java/com/bondarenko/dto/user/AuthenticationRequest.java:30]: toSting

File: src/main/java/com/bondarenko/entity/User.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    @Column(name = "name")
+    private String name;
+
+    @Column(name = "email")
+    private String email;
+
+    @Column(name = "password")
+    private String password;
+    @Transient
+    transient private String confirmPassword;
```
Comment (chickitosik) [inline src/main/java/com/bondarenko/entity/User.java:28]: move to dto, without transient

File: src/main/java/com/bondarenko/exception/user/InvalidUserValidationRequestException.java [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package com.bondarenko.exception.user;
+
+public class InvalidUserValidationRequestException extends Throwable {
+
+    public InvalidUserValidationRequestException(String name, String email) {
+      super("User is not validation, check name " + name + " email " + email );
```
Comment (chickitosik) [inline src/main/java/com/bondarenko/exception/user/InvalidUserValidationRequestException.java:6]: user credentials are not valid. Check name or email -> better for response

File: src/main/java/com/bondarenko/repo/UserRepository.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+import com.bondarenko.entity.User;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.stereotype.Repository;
+
+import java.util.Optional;
+
+@Repository
+public interface UserRepository extends JpaRepository<User, Long> {
+    @Query("select name from Users ")
```
Comment (chickitosik) [inline src/main/java/com/bondarenko/repo/UserRepository.java:12]: ( ͡° ͜ʖ ͡°)

File: src/main/java/com/bondarenko/service/UserService.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.ArrayList;
+
+
+@Slf4j
+@Service
+public class UserService implements UserDetailsService {
+    private final UserRepository userRepository;
+    // private BCryptPasswordEncoder passwordEncoder;
```
Comment (chickitosik) [inline src/main/java/com/bondarenko/service/UserService.java:21]: BCryptPasswordEncoder uncomment and use

File: src/main/java/com/bondarenko/service/UserService.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        log.info("loadUserByUsername:: Got user for " + user);
+
+        return getUserDetailsFromUser(user);
+    }
+
+    public void registration(User newUser) {
+        // passwordEncoder.encode(newUser.getPassword());
+        User registeredUser = new User(newUser.getName(), newUser.getEmail(), newUser.getPassword());
+
+        log.info("IN register - newUser: {} successfully registered", registeredUser);
```
Comment (chickitosik) [inline src/main/java/com/bondarenko/service/UserService.java:42]: do not log user password

File: src/main/java/com/bondarenko/validator/UserValidator.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    //TODO exHandler
+    @SneakyThrows
+    public void validation(String user, String email, String password) {
+        if (!userValidation(user) || !emailValidation(email) || !passwordValidation(password)) {
+            throw new InvalidUserValidationRequestException(user, email);
+        }
+    }
+
+    boolean emailValidation(String email) {
+        String emailRegex = "^[a-zA-Z0-9_+&*-]+(?:\\." +
```
Comment (chickitosik) [inline src/main/java/com/bondarenko/validator/UserValidator.java:26]: private static final pattern

File: src/main/java/com/bondarenko/validator/UserValidator.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+//            return false;//return required
+//        }
+        return pattern.matcher(userName).matches();
+    }
+
+    @SneakyThrows
+    boolean passwordValidation(String password) {
+        boolean valid = true;
+        if (password.length() > 16 || password.length() < 8) {
+            valid = false;//return required
```
Comment (chickitosik) [inline src/main/java/com/bondarenko/validator/UserValidator.java:52]: return false

File: src/main/java/com/bondarenko/validator/UserValidator.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+        }
+        if (!user.getConfirmPassword().equals(user.getPassword())) {
```
Comment (chickitosik) [inline src/main/java/com/bondarenko/validator/UserValidator.java:54]: retun `user.getConfirmPassword().equals(user.getPassword())`

File: src/main/java/com/bondarenko/controller/UserController.java (hunk 1 above)
Comment (ValentynaBondarenko) [reply src/main/java/com/bondarenko/controller/UserController.java]: done

File: src/main/java/com/bondarenko/service/impl/DefaultUserService.java [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+    @Autowired
+    UserRepository userRepository;
+    @Autowired
+    RoleRepository roleRepository;
+    @Autowired
+    PasswordEncoder encoder;
+
+    @Override
+    public ResponseEntity<?> registration(SignupRequest signUpRequest) {
+        if (userRepository.existsByUsername(signUpRequest.getUsername())) {
```
Comment (chickitosik) [inline src/main/java/com/bondarenko/service/impl/DefaultUserService.java:32]: service should return entity, not `ResponseEntity`

File: src/main/java/com/bondarenko/service/impl/DefaultUserService.java [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+
+    private User createNewUser(SignupRequest signUpRequest) {
+
+        User user = new User(signUpRequest.getUsername(), signUpRequest.getEmail(),
+                encoder.encode(signUpRequest.getPassword()));
+
+        Set<String> strRoles = signUpRequest.getRole();
+        Set<Role> roles = new HashSet<>();
+
+        setRoleNewUser(strRoles, roles);
```
Comment (chickitosik) [inline src/main/java/com/bondarenko/service/impl/DefaultUserService.java:56]: for new user set role `User`

File: src/main/java/com/bondarenko/service/impl/DefaultUserService.java [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+
+    private void setRoleNewUser(Set<String> strRoles, Set<Role> roles) {
+        if (strRoles == null) {
+            Role userRole = roleRepository.findByName(ERole.GUEST)
+                    .orElseThrow(() -> new RuntimeException("Error: Role is not found."));
+            roles.add(userRole);
+        } else {
+            strRoles.forEach(role -> {
+                switch (role) {
+                    case "admin":
```
Comment (chickitosik) [inline src/main/java/com/bondarenko/service/impl/DefaultUserService.java:70]: method can be removed

File: src/main/java/com/bondarenko/service/impl/DefaultUserService.java (hunk 11 above)
Comment (chickitosik) [reply src/main/java/com/bondarenko/service/impl/DefaultUserService.java:32]: status = created

File: src/test/java/com/bondarenko/controller/UserControllerTest.java [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+
+    @Test
+    void whenRequestParameterIsInvalid_thenReturnsStatus400() throws Exception {
+        mvc.perform(post("/signup")
+                .param("param", "3"))
+                .andExpect(status().isBadRequest());
+    }
+    @Test
+    public void whenPostRequestToUsersAndValidUser_thenCorrectResponse() throws Exception {
+        MediaType textPlainUtf8 = new MediaType(MediaType.TEXT_PLAIN, Charset.forName("UTF-8"));
```
Comment (chickitosik) [inline src/test/java/com/bondarenko/controller/UserControllerTest.java:33]: use json

File: src/test/java/com/bondarenko/dto/user/request/SignupRequestValidationTest.java [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+        //then
+        assertTrue(violations1.isEmpty());
+        assertTrue(violations2.isEmpty());
+    }
+
+    //dose correct name signupRequest3 ??
+    @Test
+    public void shouldDetectInvalidEmail() {
+        //give
+        SignupRequest signupRequest3 = new SignupRequest("Jack", "user@@example.com", "Qwerty12");
```
Comment (chickitosik) [inline src/test/java/com/bondarenko/dto/user/request/SignupRequestValidationTest.java:59]: use parametrized with `String [][] -> {mail, password}` or paramtrized 1) `String [] -> {email}`, 2) `String [] -> password`

**Pull request merged**